dma: configures the support of dma memory to memory mode

FRASTM opened the pull request:
Refine the #20824 dma mem-to-mem transfer operations on STM32WB series.
Tested on nucleo_wb55rg with dma tests loop_transfer and chan_blen_transfer.

This PR prepares  the introduction of the dmamux for stm32 soc series #22021

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
fix Nits issues

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -257,14 +257,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 
-	if ((stream->direction == MEMORY_TO_MEMORY) &&
+#ifdef CONFIG_DMA_STM32_V1
+if ((stream->direction == MEMORY_TO_MEMORY) &&
```
Wrong identation

FRASTM replied:
fixed

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -257,14 +257,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
- #ifdef CONFIG_SOC_SERIES_STM32WBX
+#endif /* CONFIG_DMA_STM32_V1 */
+	
+#ifdef CONFIG_DMA_STM32_V2
```
This restricts V2C to mem2 mem operation, which is not what is advertised in commit message.

FRASTM replied:
for DMA_V2 this part is not needed as the stream direction is not set yet
and the `dev_config->support_m2m` is not supposed to be set and is just ignored

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -257,14 +257,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 		return -EINVAL;
```
Commit message is unclear compared to changes.
Something like "restrict mem2mem limitation to V1"... would be clear (if I understood the change correctly).

FRASTM replied:
changed: no restriction on peripheral to/from memory dma transfer

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -117,7 +117,7 @@ static void dma_stm32_irq_handler(void *arg)
 }
 
-static u32_t dma_stm32_width_config(struct dma_config *config,
+static int dma_stm32_width_config(struct dma_config *config,
```
This change is not documented

FRASTM replied:
like other functions in this file, `dma_stm32_width_config` cannot return `-EINVAL` as `u32_t`
return value must be typed as **int**.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -258,14 +225,14 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 
 #ifdef CONFIG_DMA_STM32_V1
-if ((stream->direction == MEMORY_TO_MEMORY) &&
+	if ((stream->direction == MEMORY_TO_MEMORY) &&
```
Move this to previous commit

FRASTM replied:
this part is in 6498dd86cf7ecf6b55f55bd2f94b77defed7c97f

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -258,14 +225,14 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 		return -ENOTSUP;
 	}
 #endif /* CONFIG_DMA_STM32_V1 */
-	
```
ditto

FRASTM replied:
this part is in 6498dd8

FRASTM committed:
dts: bindings: dma add support for controller

This patch introduces the periph to/from memory dma transfer
define new values for dma cells on client side

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
change the driver/dma/Kconfig.stm32 to remove the dependency on HEAP_MEM_POOL_SIZE

FRASTM commented:
- refine comment  in commit acaf499f49441404b6cf2d1ab47cff50002cc339
- include the HEAP_MEM_POOL_SIZE  direcly in the Kconfig.stm32 instead of 
	default 1024

FRASTM committed:
soc: arm: stm32: define the heap size when enabling dma feature

when DMA IP is enabled on stm32 series,
the heap size must be configured to large value
because of dma channels configuration.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Kconfig: dma/Kconfig.stm32

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -605,14 +605,17 @@ static void dma_stm32_config_irq_0(struct device *dev)
-#endif
-#endif
-/* Either 5 or 7 or 8 channels for DMA1 across all stm32 series. */
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_5 */
```
DT_INST_**0**_ST_STM32_DMA_IRQ_5

FRASTM replied:
changing comment

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -629,12 +632,14 @@ static void dma_stm32_config_irq_1(struct device *dev)
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_5 */
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_6 */
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_7 */
+/* Either 0 or 5 or 6 or 7 or 8 channels for DMA2 across all stm32 series. */
```
Since instances are permutable, we should support the same #idef scheme on both instances

FRASTM replied:
ok, updating comment

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -113,6 +113,9 @@ if DMA
 config DMA_STM32
 	default y
 
+config HEAP_MEM_POOL_SIZE
```
Putting this directly in `drivers/dma/Kconfig.stm32 ` would help finding the info faster.

erwango replied:
Since symbol has to be in a defconfig this file is a good choice.

FRASTM committed:
drivers: dma: wrong return type of dma_stm32_width_config function

The dma_stm32_width_config must return a int type
as -EINVAL is a possible return value

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
this PR is to enable dma controller on the stm32 series, but does not configure to a soc

erwango commented on dts/arm/st/wb/stm32wb.dtsi:
```diff
@@ -278,6 +278,26 @@
+			label = "DMA_2";
+		};
+
 		lptim1: timers@40007c00 {
```
Please remove this

FRASTM replied:
removed

FRASTM committed:
dts: bindings: dma controller supports mem2mem transfer

This checks if the DMA controller supports or not
the memory-to-memory transfers. For DMA Version1,
in the stm32f2xx, stm32f4xx, stm32f7xx series,
only DMA instance 2 is able to transfer mem-to-mem.
For other series, with DMA Version2, there is no such a limitation.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
driver: dma: update for different dma channels

This patches defines constants from dma registers
depending on the dma configuration of the stm32 soc
Some devices have 6 or 7 or 8 dma channels per dma instance

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase

FRASTM committed:
driver: dma: split header file differently

This patches just changes the header file organisation

Signed-off-by: Francois Ramu <francois.ramu@st.com>
